### PR TITLE
[Caffe2] MIOpen bug fixes and performance enhancements

### PR DESCRIPTION
**Pull request by ashishfarmer**
This PR contains changes for:
1. Performance enhancements for group conv using MIOpen
2. Performance enhancements by removing unnecessary computations while running pooling through MIOpen
3. Added check for bwdData comptutation while running MIOpen convGradient operator 
4. Fix in MIOpen poolingGradient operator to compute window size for global pooling case
5. Minor code cleanup in MIOpen spatial batch norm operator

cc: @bddppq @petrex

### CODE REVIEW

**Comment by bddppq** [comment]
@petrex @ashishfarmer yes you can trigger the tests here https://ci.pytorch.org/jenkins/job/caffe2-builds/job/py2-clang3.8-rocm1.7.1-ubuntu16.04-trigger-test/build?delay=0sec with "GIT_COMMIT" as "origin/pr/YOUR_PR_NUMBER/head" (in this case it's "origin/pr/11766/head").

Triggered here: https://ci.pytorch.org/jenkins/job/caffe2-builds/job/py2-clang3.8-rocm1.7.1-ubuntu16.04-trigger-test/13272/

**Review by bddppq — COMMENTED** [review]
LG. Could you add a test case that could trigger the issue that you fixed?

File: caffe2/operators/hip/pool_op_miopen.cc [hunk 1]
```diff
@@ -27,7 +27,7 @@ class MIOPENPoolOp : public ConvPoolOpBase<HIPContext> {
         alpha_(OperatorBase::GetSingleArgument<float>("alpha", 1.0)),
         beta_(OperatorBase::GetSingleArgument<float>("beta", 0.0)),
         do_backward_(
-            OperatorBase::GetSingleArgument<bool>("do_backward", true)),
+            OperatorBase::GetSingleArgument<bool>("do_backward", false)),
```
Comment (bddppq) [inline caffe2/operators/hip/pool_op_miopen.cc]: I don't see this argument getting set anywhere, we can probably even remove it?

**Comment by bddppq** [comment]
https://github.com/pytorch/pytorch/blob/5392b12fa886f101fc8789cdb3db2bb6e722dd50/caffe2/operators/hip/spatial_batch_norm_op_miopen.cc#L143-L144
and
https://github.com/pytorch/pytorch/blob/5392b12fa886f101fc8789cdb3db2bb6e722dd50/caffe2/operators/hip/spatial_batch_norm_op_miopen.cc#L278-L279
also look unnecessary to me.

**Comment by bddppq** [comment]
And https://github.com/pytorch/pytorch/blob/5392b12fa886f101fc8789cdb3db2bb6e722dd50/caffe2/operators/hip/spatial_batch_norm_op_miopen.cc#L280 needs to update miopen_input_dims_

**Comment by ashishfarmer** [comment]
The case that is fixed in this PR (global pool gradient) is already covered by test_global_pooling in caffe2/python/operator_test/pooling_test.py

**Comment by bddppq** [comment]
@ashishfarmer lol then why was it not failing before?

**Comment by ashishfarmer** [comment]
It was an intermittent failure before, that we got to the root cause of with this uninitialized variable issue. It is a very specific case (MIOpen engine + AveragePool + NCHW layout) when it will be triggered, and hypothesis would for the most part not encounter it.

File: caffe2/operators/hip/pool_op_miopen.cc (hunk 1 above)
Comment (ashishfarmer) [reply caffe2/operators/hip/pool_op_miopen.cc]: Yes. This is unused. Removed it

**Comment by ashishfarmer** [comment]
> Also
> [pytorch/caffe2/operators/hip/spatial_batch_norm_op_miopen.cc](https://github.com/pytorch/pytorch/blob/5392b12fa886f101fc8789cdb3db2bb6e722dd50/caffe2/operators/hip/spatial_batch_norm_op_miopen.cc#L143-L144)
> 
> Lines 143 to 144 in [5392b12](/pytorch/pytorch/commit/5392b12fa886f101fc8789cdb3db2bb6e722dd50)
> 
>  vector<int> dims = {N, C, H, W, D}; 
>  vector<int> strides = {C * H * W * D, H * W * D, W * D, D, 1}; 
> 
> and
> [pytorch/caffe2/operators/hip/spatial_batch_norm_op_miopen.cc](https://github.com/pytorch/pytorch/blob/5392b12fa886f101fc8789cdb3db2bb6e722dd50/caffe2/operators/hip/spatial_batch_norm_op_miopen.cc#L278-L279)
> 
> Lines 278 to 279 in [5392b12](/pytorch/pytorch/commit/5392b12fa886f101fc8789cdb3db2bb6e722dd50)
> 
>  vector<int> dims = {N, C, H, W, D}; 
>  vector<int> strides = {C * H * W * D, H * W * D, W * D, D, 1}; 
> look unnecessary to me.

Removed these vectors

**Comment by ashishfarmer** [comment]
> And here
> 
> [pytorch/caffe2/operators/hip/spatial_batch_norm_op_miopen.cc](https://github.com/pytorch/pytorch/blob/5392b12fa886f101fc8789cdb3db2bb6e722dd50/caffe2/operators/hip/spatial_batch_norm_op_miopen.cc#L280)
> 
> Line 280 in [5392b12](/pytorch/pytorch/commit/5392b12fa886f101fc8789cdb3db2bb6e722dd50)
> 
>  MIOPEN_ENFORCE(miopenSet4dTensorDescriptor( 
> needs to update miopen_input_dims_

Good catch! Thank you! Fixed this

**Comment by ashishfarmer** [comment]
@bddppq - Added the fixes as per your review

**Review by bddppq — APPROVED** [review]

**Comment by bddppq** [comment]
Merged to master in https://github.com/pytorch/pytorch/commit/c7751f4df007e50842c4758b80a5357d42465022
